File: GFG/Python/2025-08-01_Balancing_Consonants_and_Vowels_Ratio.py
# GFG POTD - Balancing Consonants and Vowels Ratio
# Date: 2025-08-01

from typing import List
from collections import defaultdict

# ************************ Brute Force ************************
class BruteSolution:
    def isBalanced(self, s: str) -> bool:
        vowels = set('aeiou')
        vow = sum(1 for c in s if c in vowels)
        return vow == (len(s) - vow)

    def countBalanced(self, arr: List[str]) -> int:
        n = len(arr)
        res = 0
        for i in range(n):
            for j in range(i, n):
                curr = "".join(arr[i:j+1])
                if self.isBalanced(curr):
                    res += 1
        return res


# ************************ Better ************************
class BetterSolution:
    def isBalanced(self, s: str) -> bool:
        vowels = set('aeiou')
        vow = sum(1 for c in s if c in vowels)
        return vow == (len(s) - vow)

    def countBalanced(self, arr: List[str]) -> int:
        n = len(arr)
        res = 0
        for i in range(n):
            curr = ""
            for j in range(i, n):
                curr += arr[j]
                if self.isBalanced(curr):
                    res += 1
        return res


# ************************ Optimal (Prefix Balance + defaultdict) ************************
class OptimalSolution1:
    def countBalanced(self, arr: List[str]) -> int:
        mp = defaultdict(int)
        mp[0] = 1
        bal = 0
        res = 0
        for word in arr:
            for c in word:
                if c in 'aeiou':
                    bal += 1
                else:
                    bal -= 1
            res += mp[bal]
            mp[bal] += 1
        return res


# ************************ Optimal (Manual HashMap) ************************
class OptimalSolution2:
    def countBalanced(self, arr: List[str]) -> int:
        mp = {0: 1}
        bal = 0
        res = 0
        for word in arr:
            for c in word:
                if c in 'aeiou':
                    bal += 1
                else:
                    bal -= 1
            res += mp.get(bal, 0)
            mp[bal] = mp.get(bal, 0) + 1
        return res

File: GFG/C++/2025-08-01_Balancing_Consonants_and_Vowels_Ratio.cpp
// GFG POTD - Balancing Consonants and Vowels Ratio
// Date: 2025-08-01

#include <bits/stdc++.h>
using namespace std;

// ************************ Helper ************************
bool isVowel(char c) {
    return (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u');
}

// ************************ Brute Force ************************
class BruteSolution {
public:
    bool isBalanced(string &s) {
        int vow = 0;
        for (char c : s)
            if (isVowel(c)) vow++;
        int cons = s.length() - vow;
        return vow == cons;
    }

    int countBalanced(vector<string> &arr) {
        int n = arr.size(), res = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                string curr = "";
                for (int k = i; k <= j; k++) curr += arr[k];
                if (isBalanced(curr)) res++;
            }
        }
        return res;
    }
};

// ************************ Better (String Building On-the-Go) ************************
class BetterSolution {
public:
    bool isBalanced(string &s) {
        int vow = 0;
        for (char c : s)
            if (isVowel(c)) vow++;
        return vow == (s.length() - vow);
    }

    int countBalanced(vector<string> &arr) {
        int n = arr.size(), res = 0;
        for (int i = 0; i < n; i++) {
            string curr = "";
            for (int j = i; j < n; j++) {
                curr += arr[j];
                if (isBalanced(curr)) res++;
            }
        }
        return res;
    }
};

// ************************ Optimal (Prefix Sum + HashMap) ************************
class OptimalSolution {
public:
    int countBalanced(vector<string> &arr) {
        unordered_map<int, int> freq;
        freq[0] = 1;

        int balance = 0, res = 0;

        for (string &s : arr) {
            for (char c : s) {
                if (isVowel(c)) balance++;
                else balance--;
            }
            res += freq[balance];
            freq[balance]++;
        }

        return res;
    }
};
